streams/src/main/java/org/apache/kafka/streams/processor/internals/KafkaStreamsInstances.java

```diff
@@ -0,0 +1,227 @@
+    private final List<KafkaStreamsInstance> allInstances = new ArrayList<>();
+    private Map<String, List<TopicPartition>> partitionsByTopic = new HashMap<>();
+
+    public KafkaStreamsInstances(final TopologyBuilder builder) {
+        this.builder = builder;
+    }
+
+    /**
+     * Find all of the {@link KafkaStreamsInstance}s in a
+     * {@link KafkaStreams application}
+     *
+     * @return all the {@link KafkaStreamsInstance}s in a {@link KafkaStreams} application
+     */
+    public synchronized Collection<KafkaStreamsInstance> getAllStreamsInstances() {
+        return allInstances;
+    }
+    
+    /**
+     * Find all of the {@link KafkaStreamsInstance}s for a given storeName
+     *
+     * @param storeName the storeName to find metadata for
+     * @return A collection of {@link KafkaStreamsInstance} that have the provided storeName
+     */
+    public synchronized Collection<KafkaStreamsInstance> getAllStreamsInstancesWithStore(final String storeName) {
+        if (storeName == null) {
+            throw new IllegalArgumentException("storeName cannot be null");
+        }
+        final Set<String> sourceTopics = builder.stateStoreNameToSourceTopics().get(storeName);
+        if (sourceTopics == null) {
+            return Collections.emptyList();
+        }
+
+        final ArrayList<KafkaStreamsInstance> results = new ArrayList<>();
+        for (KafkaStreamsInstance instance : allInstances) {
+            if (instance.stateStoreNames().contains(storeName)) {
+                results.add(instance);
+            }
+        }
+        return results;
+    }
+
+    /**
+     * Find the {@link KafkaStreamsInstance}s for a given storeName and key.
+     * Note: the key may not exist in the {@link org.apache.kafka.streams.processor.StateStore},
+     * this method provides a way of finding which {@link KafkaStreamsInstance} it would exist on.
+     *
+     * @param storeName     Name of the store
+     * @param key           Key to use to for partition
+     * @param keySerializer Serializer for the key
+     * @param <K>           key type
+     * @return The {@link KafkaStreamsInstance} for the storeName and key
+     */
+    public synchronized <K> KafkaStreamsInstance getStreamsInstanceWithKey(final String storeName,
+                                                              final K key,
+                                                              final Serializer<K> keySerializer) {
+        if (keySerializer == null) {
+            throw new IllegalArgumentException("keySerializer cannot be null");
+        }
+
+        // To find the correct instance for a given key we need to use the same partitioning
+        // strategy that was used when producing the messages. In this case we create a StreamPartitioner
+        // that works the same as org.apache.kafka.clients.producer.internals.DefaultPartitioner
+        return getStreamsInstanceWithKey(storeName, key, new StreamPartitioner<K, Object>() {
```

Comment:
Correct - the user needs to know how the data has been partitioned. That is why there is an overloaded version of the method where the user supplies the `StreamPartitioner`